proposed/logging-generator.md

```diff
@@ -0,0 +1,308 @@
+# Logging Generator for strongly-typed logging messages
+
+## Summary
+
+This is a proposal for a new approach to logging to allow for more easier setup of strongly-typed logging messages. This can be achieved by using a source generator, that gets triggered with a `LoggerMessageAttribute` on partial logging methods, and would be able to either autogenerate the implementation of such method, or produce compile-time diagnostics hinting developers into proper usage of this logging approach.
+
+## Scenarios
+
+Heads up that `LoggingSample1` and `LoggingSample2` show existing approaches to logging, and the remaining will illustrate the samples using the source generator approach with `LoggerMessageAttribute`.
+
+## Background / Existing Approaches
+
+Currently there are two forms possible for doing logging, using `Microsoft.Extensions.Logging`:
+
+```c#
+public class LoggingSample1
+{
+    private ILogger _logger;
+
+    public LoggingSample1(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    public void LogMethod(string name)
+    {
+        _logger.LogInformation("Hello {name}", name);
+    }
+}
+```
+
+Here are some problems with the `LoggingSample1` sample using `LogInformation`, `LogWarning`, etc.: 
+1. We cannot provide event ID through these APIs
+2. Parameters passed are processed before LogLevel checks
+3. It requires parsing of message string on every use to find templates to substitute
+
+Because of these problems, the more efficient approach recommended today as best practices is to use [LoggerMessage.Define](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/loggermessage?view=aspnetcore-5.0) APIs instead, illustrated below with `LoggingSample2`:
+
+```c#
+public class LoggingSample2
+{
+    private ILogger _logger;
+
+    public LoggingSample2(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    public void LogMethod(string name)
+    {
+        Log.LogName(_logger, name);
+    }
+
+    private static class Log
+    {
+        private static readonly Action<ILogger, string, Exception> _logName = LoggerMessage.Define<string>(LogLevel.Information, 0, @"Hello {name}");
+
+        public static void LogName(ILogger logger, string name)
```

Comment:
I would assume this to be an extension method, but it is currently missing the `this` keyword? Is that intentional?

Reply:
It's not required to be one. I also don't think partial extension methods are supported today

Reply:
The logging method can be:

* A plain static method.
* An extension method.
* An instance method (where the ILogger comes from a field on the containing type).